File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/ArenaManager.java
package com.gmail.borlandlp.minigamesdtools.arena;

import java.util.*;
import java.util.logging.Level;
import java.util.stream.Collectors;

import com.gmail.borlandlp.minigamesdtools.APIComponent;
import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyAddedException;
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyInitializedException;
import com.gmail.borlandlp.minigamesdtools.arena.localevent.ArenaPlayerJoinLocalEvent;
import com.gmail.borlandlp.minigamesdtools.arena.localevent.ArenaPlayerLeaveLocalEvent;
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamController;
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamProvider;
import com.gmail.borlandlp.minigamesdtools.conditions.PlayerCheckResult;
import com.gmail.borlandlp.minigamesdtools.config.ConfigEntity;
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath;
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider;
import com.gmail.borlandlp.minigamesdtools.events.ArenaPlayerEnterEvent;
import com.gmail.borlandlp.minigamesdtools.events.ArenaPlayerQuitEvent;
import com.gmail.borlandlp.minigamesdtools.party.Party;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

public class ArenaManager implements APIComponent, ArenaAPI {
    private Map<String, ArenaBase> mArenas = new Hashtable<>();
    private List<String> reg2Load = new ArrayList<>();

    @Override
    public void onLoad() {
        ArenaBase arena = null;
        for(ConfigEntity configEntity : MinigamesDTools.Companion.getInstance().getConfigProvider().getPoolContents(ConfigPath.ARENA_FOLDER)) {
            this.registerToLoad(configEntity.getID());
        }

        for (String arenaName : this.reg2Load) {
            try {
                arena = MinigamesDTools.Companion.getInstance().getArenaCreatorHub().createArena(arenaName, new DataProvider());
                try {
                    arena.initializeComponents();
                } catch (ArenaAlreadyInitializedException e) {
                    e.printStackTrace();
                    continue;
                }
                this.addArena(arena);
                Debug.print(Debug.LEVEL.NOTICE, "successful load arena '" + arenaName + "'");
            } catch(Exception ex) {
                MinigamesDTools.Companion.getInstance().getLogger().log(Level.WARNING, " failed to load arena " + arenaName);
                ex.printStackTrace();
            }
        }
    }

    @Override
    public void onUnload() {
        for(ArenaBase arena : this.getEnabledArenas()) {
            arena.forceDisable();
        }
    }

    @Override
    public void registerToLoad(String arenaName) {
        this.reg2Load.add(arenaName);
    }

    @Override
    public void unregisterToLoad(String arenaName) {
        this.reg2Load.remove(arenaName);
    }

    @Override
    public void restartArena(String name) {
        if(!this.mArenas.containsKey(name)) {
            return;
        } else if(this.mArenas.get(name).getState() != ArenaBase.STATE.ENDED) {
            this.mArenas.get(name).gameEnded(false);
        }

        ArenaBase oldInstance = this.mArenas.get(name);
        this.mArenas.remove(name);
        ArenaBase newInstance = null;
        try {
            newInstance = MinigamesDTools.Companion.getInstance().getArenaCreatorHub().createArena(name, new DataProvider());
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
        try {
            newInstance.initializeComponents();
        } catch (ArenaAlreadyInitializedException e) {
            e.printStackTrace();
            return;
        }
        this.mArenas.put(name, newInstance);
        Debug.print(Debug.LEVEL.NOTICE, "Reload arena '" + name + "', oldInstance:" + oldInstance + " # newInstance:" + newInstance);
    }

    @Override
    public void addArena(ArenaBase arena) throws ArenaAlreadyAddedException {
        for(ArenaBase qArena : this.getArenas()) {
            if(qArena.getName().equalsIgnoreCase(arena.getName())) {
                String msg = "Arena[name:" + arena.getName() + "] with a similar name is already registered.";
                throw new ArenaAlreadyAddedException(msg);
            }
        }

        this.mArenas.put(arena.getName(), arena);
    }

    @Override
    public void removeArena(String arenaName) {
        this.mArenas.remove(arenaName);
    }

    public ArenaBase getArenaOf(Player player) {
        for (ArenaBase arenaBase : this.getEnabledArenas()) {
            if(arenaBase.getTeamController().getTeamOf(player) != null) return arenaBase;
        }

        return null;
    }

    @Override
    public ArenaBase getArenaOf(String playerName) {
        for (ArenaBase arenaBase : this.getEnabledArenas()) {
            if(arenaBase.getTeamController().getTeamOf(playerName) != null) return arenaBase;
        }

        return null;
    }

    public ArenaBase getArena(String arenaName) {
       return this.mArenas.get(arenaName);
   }

    public ArrayList<ArenaBase> getArenas() {
      return new ArrayList<>(this.mArenas.values());
   }

    @Override
    public ArrayList<ArenaBase> getEnabledArenas() {
        return this.getArenas().stream().filter(ArenaBase::isEnabled).collect(Collectors.toCollection(ArrayList::new));
    }

    @Override
    public ArrayList<ArenaBase> getDisabledArenas() {
        return this.getArenas().stream().filter(x -> !x.isEnabled()).collect(Collectors.toCollection(ArrayList::new));
    }

    @Override
    public void arenaLeaveRequest(Player player) {
        ArenaBase arena = this.getArenaOf(player);
        if(arena != null) {
            arena.getEventAnnouncer().announce(new ArenaPlayerLeaveLocalEvent(player));
        } else {
            Debug.print(Debug.LEVEL.NOTICE, "Cant find arena for Player[name:" + player.getName() + "]");
            return;
        }

        ArenaPlayerQuitEvent event = new ArenaPlayerQuitEvent(arena, player);
        Bukkit.getPluginManager().callEvent(event);

        if(arena.getState() == ArenaBase.STATE.DELAYED_START) {
            if(arena.getTeamController().countCurrentPlayers() < arena.getGameRules().minPlayersToStart) {
                arena.cancelDelayedStartArena();
            }
        }
    }

    @Override
    public boolean arenaJoinRequest(String arenaName, Party party) {
        ArenaBase arenaBase = this.getArena(arenaName);
        if(arenaBase == null) {
            party.getLeader().sendMessage(MinigamesDTools.Companion.getPrefix() + " Arena " + arenaName + " does not exist.");
            return false;
        }

        TeamController teamControl = arenaBase.getTeamController();
        TeamProvider team = teamControl.getTeams().stream()
                .filter(t -> t.containsFreeSlots(party.getPlayers().size()))
                .findFirst()
                .orElse(null);
        if(team == null) {
            party.getLeader().sendMessage("Arena " + arenaName + " is full");
            return false;
        }

        PlayerCheckResult result;
        for (Player player : party.getPlayers()) {
            result = this.playerCanJoin(arenaBase, player);
            if(result.getResult() == PlayerCheckResult.CheckResult.DENIED) {
                String msg = "A member of your party(" + player.getName() + ") does not meet the conditions of the gaming arena for the following reasons.";
                party.getLeader().sendMessage(msg);
                result.getErrId().forEach(party.getLeader()::sendMessage);
                return false;
            }
        }

        for (Player player : party.getPlayers()) {
            arenaBase.getEventAnnouncer().announce(new ArenaPlayerJoinLocalEvent(player, team));
            Bukkit.getPluginManager().callEvent(new ArenaPlayerEnterEvent(arenaBase, player));
        }

        if(arenaBase.getTeamController().countCurrentPlayers() >= arenaBase.getGameRules().minPlayersToStart && arenaBase.getState() == ArenaBase.STATE.EMPTY) {
            arenaBase.delayedStartArena();
        }

        return true;
    }

    @Override
    public boolean arenaJoinRequest(String arenaName, Player player) {
        ArenaBase arenaBase = this.getArena(arenaName);
        if(arenaBase == null) {
            player.sendMessage(MinigamesDTools.Companion.getPrefix() + " Arena " + arenaName + " does not exist.");
            return false;
        }

        TeamController teamControl = arenaBase.getTeamController();
        TeamProvider team = teamControl.getTeams().stream().filter(t -> t.containsFreeSlots(1)).findFirst().orElse(null);
        if(team == null) {
            player.sendMessage("Arena " + arenaName + " is full");
            return false;
        }

        // TODO: Приделать конвертацию ID в Localizer
        PlayerCheckResult result = this.playerCanJoin(arenaBase, player);
        if(result.getResult() == PlayerCheckResult.CheckResult.DENIED) {
            result.getErrId().forEach(player::sendMessage);
            return false;
        }

        arenaBase.getEventAnnouncer().announce(new ArenaPlayerJoinLocalEvent(player, team));
        Bukkit.getPluginManager().callEvent(new ArenaPlayerEnterEvent(arenaBase, player));

        if(arenaBase.getTeamController().countCurrentPlayers() >= arenaBase.getGameRules().minPlayersToStart && arenaBase.getState() == ArenaBase.STATE.EMPTY) {
            arenaBase.delayedStartArena();
        }

        return true;
    }

   @Override
   public void disableArena(String arenaName) {
       ArenaBase arena = this.getArena(arenaName);
       if(arena != null) {
          arena.setEnabled(false);
          Bukkit.getServer().broadcastMessage("[MinigamesDTools] Выключена арена " + arenaName);
       }
   }

   @Override
   public void enableArena(String arenaName) {
        ArenaBase arena = this.getArena(arenaName);
        if(arena != null) {
            arena.setEnabled(true);
            Bukkit.getServer().broadcastMessage("[MinigamesDTools] Запущена арена " + arenaName);
        }
   }

    @Override
    public PlayerCheckResult playerCanJoin(ArenaBase arenaBase, Player player) {
        return arenaBase.getJoinConditionsChain().check(player);
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/ArenaManager.kt
package com.gmail.borlandlp.minigamesdtools.arena

import com.gmail.borlandlp.minigamesdtools.APIComponent
import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools.Companion.instance
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools.Companion.prefix
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyAddedException
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyInitializedException
import com.gmail.borlandlp.minigamesdtools.arena.localevent.ArenaPlayerJoinLocalEvent
import com.gmail.borlandlp.minigamesdtools.arena.localevent.ArenaPlayerLeaveLocalEvent
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamProvider
import com.gmail.borlandlp.minigamesdtools.conditions.PlayerCheckResult
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider
import com.gmail.borlandlp.minigamesdtools.events.ArenaPlayerEnterEvent
import com.gmail.borlandlp.minigamesdtools.events.ArenaPlayerQuitEvent
import com.gmail.borlandlp.minigamesdtools.party.Party
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*
import java.util.function.Consumer
import java.util.logging.Level
import kotlin.collections.MutableList
import kotlin.collections.MutableMap
import kotlin.collections.set

class ArenaManager : APIComponent, ArenaAPI {
    private val mArenas: MutableMap<String, ArenaBase> = Hashtable()
    private val reg2Load: MutableList<String> = mutableListOf()

    override fun onLoad() {
        var arena: ArenaBase?
        for (configEntity in instance!!.configProvider!!.getPoolContents(ConfigPath.ARENA_FOLDER)) {
            registerToLoad(configEntity.id)
        }
        for (arenaName in reg2Load) {
            try {
                print(Debug.LEVEL.NOTICE, "Start load arena '$arenaName'")
                arena = instance!!.arenaCreatorHub!!.createArena(arenaName, DataProvider())
                try {
                    arena.initializeComponents()
                } catch (e: ArenaAlreadyInitializedException) {
                    e.printStackTrace()
                    continue
                }
                addArena(arena)
                print(Debug.LEVEL.NOTICE, "successful load arena '$arenaName'")
            } catch (ex: Exception) {
                instance!!.logger.log(Level.WARNING, " failed to load arena $arenaName")
                ex.printStackTrace()
            }
        }
    }

    override fun onUnload() {
        for (arena in enabledArenas) {
            arena.forceDisable()
        }
    }

    override fun registerToLoad(arenaName: String) {
        reg2Load.add(arenaName)
    }

    override fun unregisterToLoad(arenaName: String) {
        reg2Load.remove(arenaName)
    }

    override fun restartArena(arenaName: String) {
        if (!mArenas.containsKey(arenaName)) {
            return
        } else if (mArenas[arenaName]!!.state != ArenaBase.STATE.ENDED) {
            mArenas[arenaName]!!.gameEnded(false)
        }
        val oldInstance = mArenas[arenaName]
        mArenas.remove(arenaName)

        val newInstance: ArenaBase = try {
            instance!!.arenaCreatorHub!!.createArena(arenaName, DataProvider())
        } catch (e: Exception) {
            e.printStackTrace()
            return
        }
        try {
            newInstance.initializeComponents()
        } catch (e: ArenaAlreadyInitializedException) {
            e.printStackTrace()
            return
        }
        mArenas[arenaName] = newInstance
        print(
            Debug.LEVEL.NOTICE,
            "Reload arena '$arenaName', oldInstance:$oldInstance # newInstance:$newInstance"
        )
    }

    @Throws(ArenaAlreadyAddedException::class)
    override fun addArena(arena: ArenaBase) {
        for (qArena in arenas) {
            if (qArena.name.equals(arena.name, ignoreCase = true)) {
                val msg =
                    "Arena[name:" + arena.name + "] with a similar name is already registered."
                throw ArenaAlreadyAddedException(msg)
            }
        }
        mArenas[arena.name!!] = arena
    }

    override fun removeArena(arenaName: String) {
        mArenas.remove(arenaName)
    }

    override fun getArenaOf(player: Player): ArenaBase? {
        for (arenaBase in enabledArenas) {
            if (arenaBase.teamController!!.getTeamOf(player) != null) return arenaBase
        }
        return null
    }

    override fun getArenaOf(playerName: String): ArenaBase? {
        for (arenaBase in enabledArenas) {
            if (arenaBase.teamController!!.getTeamOf(playerName) != null) return arenaBase
        }
        return null
    }

    override fun getArena(arenaName: String): ArenaBase? {
        return mArenas[arenaName]
    }

    override val arenas: List<ArenaBase>
        get() = ArrayList(mArenas.values)

    override val enabledArenas: List<ArenaBase>
        get() = arenas.filter { obj: ArenaBase -> obj.isEnabled }

    override val disabledArenas: List<ArenaBase>
        get() = arenas.filter { x: ArenaBase -> !x.isEnabled }

    override fun arenaLeaveRequest(player: Player) {
        val arena = this.getArenaOf(player)
        if (arena != null) {
            arena.eventAnnouncer.announce(ArenaPlayerLeaveLocalEvent(player))
        } else {
            print(
                Debug.LEVEL.NOTICE,
                "Cant find arena for Player[name:" + player.name + "]"
            )
            return
        }
        val event = ArenaPlayerQuitEvent(arena, player)
        Bukkit.getPluginManager().callEvent(event)
        if (arena.state == ArenaBase.STATE.DELAYED_START) {
            if (arena.teamController!!.countCurrentPlayers() < arena.gameRules!!.minPlayersToStart) {
                arena.cancelDelayedStartArena()
            }
        }
    }

    override fun arenaJoinRequest(arenaName: String, party: Party): Boolean {
        val arenaBase = getArena(arenaName)
        if (arenaBase == null) {
            party.leader.sendMessage("$prefix Arena $arenaName does not exist.")
            return false
        }
        val teamControl = arenaBase.teamController
        val team = teamControl!!.teams.stream()
            .filter { t: TeamProvider -> t.containsFreeSlots(party.players.size) }
            .findFirst()
            .orElse(null)
        if (team == null) {
            party.leader.sendMessage("Arena $arenaName is full")
            return false
        }
        var result: PlayerCheckResult
        for (player in party.players) {
            result = playerCanJoin(arenaBase, player)
            if (result.result === PlayerCheckResult.CheckResult.DENIED) {
                val msg =
                    "A member of your party(" + player.name + ") does not meet the conditions of the gaming arena for the following reasons."
                party.leader.sendMessage(msg)
                result.errId.forEach(Consumer { s: String? ->
                    party.leader.sendMessage(
                        s
                    )
                })
                return false
            }
        }
        for (player in party.players) {
            arenaBase.eventAnnouncer.announce(ArenaPlayerJoinLocalEvent(player, team))
            Bukkit.getPluginManager().callEvent(ArenaPlayerEnterEvent(arenaBase, player))
        }
        if (arenaBase.teamController!!.countCurrentPlayers() >= arenaBase.gameRules!!.minPlayersToStart && arenaBase.state == ArenaBase.STATE.EMPTY) {
            arenaBase.delayedStartArena()
        }
        return true
    }

    override fun arenaJoinRequest(arenaName: String, player: Player): Boolean {
        val arenaBase = getArena(arenaName)
        if (arenaBase == null) {
            player.sendMessage("$prefix Arena $arenaName does not exist.")
            return false
        }
        val teamControl = arenaBase.teamController
        val team = teamControl!!.teams.stream()
            .filter { t: TeamProvider -> t.containsFreeSlots(1) }
            .findFirst().orElse(null)
        if (team == null) {
            player.sendMessage("Arena $arenaName is full")
            return false
        }
        // TODO: Приделать конвертацию ID в Localizer
        val result = playerCanJoin(arenaBase, player)
        if (result.result === PlayerCheckResult.CheckResult.DENIED) {
            result.errId.forEach(Consumer { s: String? -> player.sendMessage(s) })
            return false
        }
        arenaBase.eventAnnouncer.announce(ArenaPlayerJoinLocalEvent(player, team))
        Bukkit.getPluginManager().callEvent(ArenaPlayerEnterEvent(arenaBase, player))
        if (arenaBase.teamController!!.countCurrentPlayers() >= arenaBase.gameRules!!.minPlayersToStart && arenaBase.state == ArenaBase.STATE.EMPTY) {
            arenaBase.delayedStartArena()
        }
        return true
    }

    override fun disableArena(arenaName: String) {
        val arena = getArena(arenaName)
        if (arena != null) {
            arena.isEnabled = false
            Bukkit.getServer().broadcastMessage("[MinigamesDTools] Выключена арена $arenaName")
        }
    }

    override fun enableArena(arenaName: String) {
        val arena = getArena(arenaName)
        if (arena != null) {
            arena.isEnabled = true
            Bukkit.getServer().broadcastMessage("[MinigamesDTools] Запущена арена $arenaName")
        }
    }

    override fun playerCanJoin(arenaBase: ArenaBase, player: Player): PlayerCheckResult {
        return arenaBase.joinConditionsChain!!.check(player)
    }
}